components/eamxx/src/physics/nudging/atmosphere_nudging.cpp

```diff
@@ -0,0 +1,264 @@
+  view_2d<Real> qv_aft("",qv_ext.extent_int(0),qv_ext.extent_int(1));
+  Kokkos::deep_copy(hw_aft,horiz_winds_ext);
+  Kokkos::deep_copy(qv_aft,qv_ext);
+  double time_step_file_d = time_step_file;
+  double w_bef = ((time_index+1)*time_step_file-time_s) / time_step_file_d;
+  double w_aft = (time_s-(time_index)*time_step_file) / time_step_file_d;
+  
+  const int num_cols = T_mid_ext.extent(0);
+  const int num_vert_packs = T_mid_ext.extent(1);
+  const auto policy = ESU::get_default_team_policy(num_cols, num_vert_packs);
+  Kokkos::parallel_for("nudging_time_interpolation", policy,
+     	       KOKKOS_LAMBDA(MemberType const& team) {
+      const int icol = team.league_rank();
+      auto T_mid_1d = ekat::subview(T_mid_ext,icol);
+      auto T_mid_bef_1d = ekat::subview(T_mid_bef,icol);
+      auto T_mid_aft_1d = ekat::subview(T_mid_aft,icol);
+      auto hw_2d = ekat::subview(horiz_winds_ext, icol);
+      auto hw_bef_2d = ekat::subview(hw_bef, icol);
+      auto hw_aft_2d = ekat::subview(hw_aft, icol);
+      auto p_mid_1d = ekat::subview(p_mid_ext,icol);
+      auto p_mid_bef_1d = ekat::subview(p_mid_bef,icol);
+      auto p_mid_aft_1d = ekat::subview(p_mid_aft,icol);
+      auto qv_1d = ekat::subview(qv_ext,icol);
+      auto qv_bef_1d = ekat::subview(qv_bef,icol);
+      auto qv_aft_1d = ekat::subview(qv_aft,icol);
+      const auto range = Kokkos::TeamThreadRange(team, num_vert_packs);
+      Kokkos::parallel_for(range, [&] (const Int & k) {
+	  T_mid_1d(k)=w_bef*T_mid_bef_1d(k) + w_aft*T_mid_aft_1d(k);
+	  p_mid_1d(k)=w_bef*p_mid_bef_1d(k) + w_aft*p_mid_aft_1d(k);
+	  hw_2d(0,k)=w_bef*hw_bef_2d(0,k) + w_aft*hw_aft_2d(0,k);
+	  hw_2d(1,k)=w_bef*hw_bef_2d(1,k) + w_aft*hw_aft_2d(1,k);
+	  qv_1d(k)=w_bef*qv_bef_1d(k) + w_aft*qv_aft_1d(k);
+      });
+      team.team_barrier();
+  });
+  Kokkos::fence();   
+
+}
+  
+// =========================================================================================
+void NUDGING::run_impl (const int dt)
+{
+  using namespace scream::vinterp;
+
+  //Have to add dt because first time iteration is at 0 seconds where you will not have
+  //any data from the field. The timestamp is only iterated at the end of the
+  //full step in scream.
+  auto ts = timestamp()+dt;
+  auto time_since_zero = ts.seconds_from(ts0);
+  //Don't nudge for times less than first time step in nudged file to avoid
+  //extrapolation
+  if (time_since_zero < time_step_file){ return;}
+  //perform time interpolation
+  time_interpolation(time_since_zero);
+
+  auto T_mid       = get_field_out("T_mid").get_view<mPack**>();
+  auto qv          = get_field_out("qv").get_view<mPack**>();
+  auto hw          = get_field_out("horiz_winds").get_view<mPack***>();
+
+  const auto& p_mid    = get_field_in("p_mid").get_view<const mPack**>();
+  auto p_mid_v = view_Nd<mPack,2>("",p_mid.extent_int(0),p_mid.extent_int(1));
+  Kokkos::deep_copy(p_mid_v,p_mid);
+  
+  view_Nd<mPack,2> T_mid_out("T_mid_out",m_num_cols,T_mid.extent_int(1));
```

Comment:
Creating views on device at runtime is not advised, due to cost of allocation. Since I see you eventually copy this back into T_mid, why don't you just use T_mid? Same goes for qv, and hw.

Reply:
As for the `_p` variables, I don't think you need them. Can't you just replace every use of, say, `T_mid_ext_p` with `T_mid_ext`?

Reply:
The reason I had to convert these is that the vertical interpolation is expecting a view_Nd, so if you try to put T_mid in directly or T_mid_ext in directly it gives the error below. Previously I think the vertical interpolation was templated so that this would have been ok, but with the recent updates to the vertical interpolation by Aaron I think you have to specify it in terms of a view_Nd. I will see if there is a way around this or if the vertical interpolation needs to be updated. 
error: no instance of overloaded function "perform_vertical_interpolation" matches the argument list
            argument types are: (scream::vinterp::view_Nd<scream::Nudging::mPack, 2>, scream::vinterp::view_Nd<scream::Nudging::mPack, 2>, scream::Nudging::view_2d<scream::Real>, scream::vinterp::view_Nd<scream::Nudging::mPack, 2>, int, int)

Reply:
I see that `T_mid_ext` is a view of Real, while `T_mid_ext_p` is a view of Packs. First, the deep copy below looks dangerous. I suspect it ony works because you are using packs of size 1.

Regardless, if you need to turn a view of Real into a view of Pack, you can just do
```
auto packed_data = reinterpret_cast<mPack*>(T_mid_ext.data());
int num_packs = ekat::PackInfo<mPack::n>::num_packs(num_levs); // <- num_levs: whatever num levs T_mid_ext uses
view_Nd<mPack,2> T_mid_ext_p ( packed_data, num_cols,num_packs); // <- num_cols: whatever num_cols T_mid_ext uses
```

Reply:
mPack always has a pack size of 1. The reason I did this was because of this issue when converting from a view of Real to a view of Pack and I thought there would be problems with a pack size different from 1 (and wasn't sure how to resolve). In the case of pack size 1 I think you could simplify this to one line(?):
view_Nd<mPack,2> T_mid_ext_p( reinterpret_cast<mPack*> (T_mid_ext.data()),num_cols,num_levs);

An alternative solution is whether it might be possible to get fields instead of views from a file using AtmosphereInput. If this is possible then I could get the fields from the external field and then just do field.get_view<mPack**>. But I am not sure if this is possible. It looks like I might be able to create a field manager and fill it with the fields and then feed it to AtmosphereInput.

Reply:
One additional comment with respect to "why don't you just use T_mid? Same goes for qv, and hw." Eventually we want to actually output to a file the nudging tendencies, i.e. |original T_mid - T_mid from external file|. So we don't want to overwrite T_mid. But this will probably come in a later PR, so for now I can just do it like you suggest with the idea that this will be updated later.

Reply:
Yes, the reintepret cast line you posted should work.

Reply:
As for AtmInput, we do not offer a way to read into fields, but we _could_ think of adding such capability. After all, the setup version that takes layouts and views could be easily called by a version that takes Fields. We may even benefit a tiny bit from this simplification in other parts of the EAMxx, like when we read geo data from IC file or topo from topofile.

Reply:
Ok, for now I will implement the reinterpret cast and we can think about if we want to add the AtmInput in the future.